src/cryptonote_basic/account.cpp

```diff
@@ -70,62 +71,81 @@ DISABLE_VS_WARNINGS(4244 4345)
     data[sizeof(base_key)] = config::HASH_KEY_MEMORY;
     crypto::generate_chacha_key(data.data(), sizeof(data), key, 1);
   }
-  //-----------------------------------------------------------------
-  static epee::wipeable_string get_key_stream(const crypto::chacha_key &base_key, const crypto::chacha_iv &iv, size_t bytes)
+  
+  // Prepare IVs and start chacha for encryption
+  void account_keys::encrypt_wrapper(const crypto::chacha_key &key, const bool all_keys)
   {
-    // derive a new key
-    crypto::chacha_key key;
-    derive_key(base_key, key);
+    // Set a fresh IV only for all-key encryption
+    if (all_keys)
+      m_encryption_iv = crypto::rand<crypto::chacha_iv>();
 
-    // chacha
-    epee::wipeable_string buffer0(std::string(bytes, '\0'));
-    epee::wipeable_string buffer1 = buffer0;
-    crypto::chacha20(buffer0.data(), buffer0.size(), key, iv, buffer1.data());
-    return buffer1;
+    // Now do the chacha
+    chacha_wrapper(key, all_keys);
   }
-  //-----------------------------------------------------------------
-  void account_keys::xor_with_key_stream(const crypto::chacha_key &key)
+
+  // Start chacha for decryption
+  void account_keys::decrypt_wrapper(const crypto::chacha_key &key, const bool all_keys)
+  {
+    chacha_wrapper(key, all_keys);
+  }
+
+  // Decrypt keys using the legacy method
+  void account_keys::decrypt_legacy(const crypto::chacha_key &key)
   {
-    // encrypt a large enough byte stream with chacha20
-    epee::wipeable_string key_stream = get_key_stream(key, m_encryption_iv, sizeof(crypto::secret_key) * (2 + m_multisig_keys.size()));
-    const char *ptr = key_stream.data();
+    // Derive domain-separated chacha key
+    crypto::chacha_key derived_key;
+    derive_key(key, derived_key);
+
+    // Build key stream
+    epee::wipeable_string temp(std::string(sizeof(crypto::secret_key)*(2 + m_multisig_keys.size()), '\0'));
+    epee::wipeable_string stream = temp;
+    crypto::chacha20(temp.data(), temp.size(), derived_key, m_encryption_iv, stream.data());
+
+    // Decrypt all keys
+    const char *ptr = stream.data();
     for (size_t i = 0; i < sizeof(crypto::secret_key); ++i)
       m_spend_secret_key.data[i] ^= *ptr++;
     for (size_t i = 0; i < sizeof(crypto::secret_key); ++i)
       m_view_secret_key.data[i] ^= *ptr++;
     for (crypto::secret_key &k: m_multisig_keys)
     {
       for (size_t i = 0; i < sizeof(crypto::secret_key); ++i)
-        k.data[i] ^= *ptr++;
+        k.data[i] ^= *ptr;;
```

Comment:
Looks as though this may be an error, which would also suggest a missing test.

Reply:
Good catch!

Reply:
It looks like the unit test wallet file doesn't currently include any multisig addresses.